spec/rspec/mocks/and_wrap_original_spec.rb

```diff
@@ -10,54 +10,68 @@ def results
+            m.call.first(10)
+          end
+        }.to change { instance.results.size }.from(100).to(10)
+      end
 
-    it "raises a name error if the method does not exist" do
-      expect {
-        allow(instance).to receive(:no_results).and_wrap_original { |m| m.call }
-        instance.no_results
-      }.to raise_error NameError
-    end
+      it "raises a name error if the method does not exist" do
+        expect {
+          target.to receive(:no_results).and_wrap_original { |m| m.call }
+          instance.no_results
+        }.to raise_error NameError
+      end
 
-    it "passes in the original method" do
-      value = nil
-      original_method = instance.method(:results)
-      allow(instance).to receive(:results).and_wrap_original { |m| value = m }
-      instance.results
-      expect(value).to eq original_method
-    end
+      it "passes in the original method" do
+        value = nil
+        original_method = instance.method(:results)
+        target.to receive(:results).and_wrap_original { |m| value = m }
+        instance.results
+        expect(value).to eq original_method
+      end
 
-    it "passes along the message arguments" do
-      values = nil
-      allow(instance).to receive(:results).and_wrap_original { |_, *args| values  = args }
-      instance.results(1, 2, 3)
-      expect(values).to eq [1, 2, 3]
-    end
+      it "passes along the message arguments" do
+        values = nil
+        target.to receive(:results).and_wrap_original { |_, *args| values  = args }
+        instance.results(1, 2, 3)
+        expect(values).to eq [1, 2, 3]
+      end
+
+      it "passes along any supplied block" do
+        value = nil
+        target.to receive(:results).and_wrap_original { |&b| value = b }
+        instance.results(&(block = proc {}))
+        expect(value).to eq block
+      end
+
+      it "ignores previous stubs" do
+        target.to receive(:results) { "results" }
+        target.to receive(:results).and_wrap_original { |m| m.call }
+        expect(instance.results).to_not eq "results"
+      end
 
-    it "passes along any supplied block" do
-      value = nil
-      allow(instance).to receive(:results).and_wrap_original { |&b| value = b }
-      instance.results(&(block = proc {}))
-      expect(value).to eq block
+      it "can be constrained by specific arguments" do
+        target.to receive(:results) { :all }
+        target.to receive(:results).with(5).and_wrap_original { |m, n| m.call.first(n) }
+        expect(instance.results 5).to eq [1,2,3,4,5]
+        expect(instance.results).to eq :all
+      end
     end
 
-    it "ignores previous stubs" do
-      allow(instance).to receive(:results) { "results" }
-      allow(instance).to receive(:results).and_wrap_original { |m| m.call }
-      expect(instance.results).to_not eq "results"
+    describe "allow(...).to receive(...).and_wrap_original" do
+      it_behaves_like "using and_wrap_original" do
+        let(:target) { allow(instance) }
```

Comment:
What do you think about naming this `allow_it` rather than `target`?  Then at the call site it reads quite nicely:

``` ruby
allow_it.to receive(...)...
# rather than:
target.to receive(...)...
```